src/Icicle/Avalanche/Statement/Simp/Eval.hs

```diff
@@ -39,18 +39,15 @@ simpEvalX' ev ty = go
         , p' <- fromMaybe p mp
         , q' <- fromMaybe q mq
         , x' <- XApp a p' q'
-        , Just (prim, as) <- takePrimApps x'
-        , Just args       <- mapM (takeValue . just go) as
-        -> case simpEvalP ev ty a prim args of
-            Just x''
-             -> return x''
-            Nothing
-             | isJust mp || isJust mq
-             -> return x'
-             | otherwise
-             -> Nothing
+        -> case takePrimApps x' of
+             Just (prim, as)
+               | Just args <- mapM takeValue as
+               -> case simpEvalP ev ty a prim args of
+                   Just x''
+                     -> return x''
+                   _ -> XApp a <$> mp <*> mq
+             _ -> XApp a <$> mp <*> mq
```

Comment:
I think you'll have to explain to me why this is faster: I can't see it. amazing